### DISCUSSION TITLE
SLS Audio tweaking based on Device Posture

### DISCUSSION THREAD

**Discussion by NP-chaonay** [linux-surface/linux-surface#860] [Show and tell]
**(Requirements)**
- SLS
- Surface kernel >=5.18.4-surface
- Pulseeffects (currently renamed as easyeffects) (you can using other method, but I don't have config file for them)

**Purpose**: see Link A, and integrate PulseEffects preset with SLS device posture sensor is the good UX.

**what I do is**
1. config PulseEffects (or your method) to have multiple preset based on every single device posture (if preset is not implemented on your selected method, then using another script to implement this instead instead, or just builtin to below script)
- - available device posture that system can detected are: ```laptop/slate/tablet/closed (lid closed)```
- - for my lazy adjustment, see [link A](https://github.com/NP-chaonay/Misc/tree/main/SurfaceLaptopStudio_LinuxAudioTweak_OnDevicePosture), welcome for incoming contribution. (if have)
2. making script to regularly read (for me, set to every <=1sec && >=0.2sec is good settings) state from ```/sys/devices/platform/MSHW0123\:00/01\:26\:01\:00\:01/state``` (```laptop/slate/tablet/closed```) and auto execute command when new state is detected
- - or using [my script](https://github.com/NP-chaonay/Misc/blob/main/SurfaceLaptopStudio_DevicePostureSensor_Daemon/README.md) 
- - (hint: when first init script, set OLD_STATE to empty, and then get NEW_STATE and do by condition and then OLD_STATE=NEW_STATE) 
- - the command to run is ```pulseeffects -l SLS_laptop``` (depend on your preset name and program name)

**(Tricks and Troubleshooting)**
- For Tricks see link A
- if sound quality is not good and seem to be cracked, try using 48khz sample rate for sample rate of laptop speaker (which is alsa hw:0,0 when headphone disconnected) instead (see [why?](https://github.com/linux-surface/linux-surface/wiki/Surface-Laptop-Studio) for more info)
  - As StollD have said for system with PipeWire as default audio server instead of PulseAudio, you might don't have to set 48kHz sample rate setting, if you using PipeWire as default audio server since it is already default thing for PipeWire
  - if it not resolved, then maybe tweaking configuration problem

**How you can contribute**
_see link A_

_Note that I have found where Windows keep EQ data, however cannot interpret its meaning of byte codes, see in link a_

**Comment by StollD** [comment]
> if sound quality is not good, try using 48khz sample rate instead (see [why?](https://github.com/linux-surface/linux-surface/wiki/Surface-Laptop-Studio)), if not resolved then maybe tweaking configuration problem

Just for reference, pipewire with its default configuration automatically resamples everything to 48khz

**Reply by NP-chaonay** [reply]
thanks I havent know of that.

however despite of any resampling/sample rate settings of app/PA/ALSA is set before passing through ALSA hw:0,0 device. If before passing to hw:0,0, it still sounds good, there are 2 possible way to pass to ALSA HW and each have different result:

- passed as 16/32bit but 44.1kHz
  - ALSA hw:0,0 report that it is supported, however using this option cause sounds to be terrible cracks (not much crack but it just like nearly-new noise-hearable vinyl)
  - On Windows, only 16-bit 48kHz supports 
  - on ```/proc/asound/PCH/codec#0``` also shown that 44.1khz and 48khz is supported
  - I think it is because bug that poor resampling of either DAC/intel audio with realtek driver
- passed as 16/32bit but 48kHz: works

so if pipewire using 48khz by default, you got 48khz in PA because of alternative-sample-rate setting of PA, but if there is any sound already play in 44.1khz (which is default sample rate settings of PA), you got 48khz from pipewire but later converted to 44.1khz by PA and pass to hw:0,0 and got poor quality sound

PS: thanks again, I have re-edited msg to more clarify that it is sample rate that to output to hw:0,0 not the sample rate of app/Pulseeffects by itself

**Reply by StollD** [reply]
Not sure if I understand you correctly. Once pipewire is involved, pulseaudio should not play any role anymore. PW is a replacement for PA (including a compatibility layer).

My point was that if you switch from PA to PW it should work fine, because PW will manage the ALSA device, and only pass 48khz audio, no matter what sampling rate you actually pass to pipewire. (Note that you should not directly play anything to the ALSA device if it is managed by PA or PW. They usually install an ALSA plugin to create a virtual device that routes everything back to PA / PW)

**Reply by NP-chaonay** [reply]
> Once pipewire is involved, pulseaudio should not play any role anymore. PW is a replacement for PA (including a compatibility layer).

As I do my research I just know that there is PipeWire sound server, I havent know this too, if I not wrong Pipewire have been an multimedia (pipeline) API, so I guess it is around 2015-2020 where it becoming audio server. So if PW is used as audio server, then PA is just compatibility layer, and PE. do not thing to PA, however on my system still using of PA (Ubuntu 22.04, I just know that 22.10 will use PW. as default, maybe I'm out of Linux trend lamo).

> My point was that if you switch from PA to PW it should work fine, because PW will manage the ALSA device, and only pass 48khz audio,

yes

> Note that you should not directly play anything to the ALSA device if it is managed by PA or PW.

yes, thanks for suggestion, but something requires some tweaks, and using ALSA directly make me discover more about Linux system, including the thing that why ALSA is not support 24bit even codec file report as 16/20/24 bit (this I havent discovered, but I know thats it is because ALSA config file). And it use less latency (I can do it too on PA, however), and for stuff that I want to using HW direct (such as audio quality experiment, because before I know that problem is on 44.1khz it maybe PA but this test shown that even HW direct got same problem)

> They usually install an ALSA plugin to create a virtual device that routes everything back to PA / PW)

yes, I have known that. it implement ALSA config file said "!default.pcm" like this and set alsa device "pulse" as default pcm. So without config overriding, it will play on pulse by default.

> if sound quality is not good, try using 48khz sample rate for sample rate of speaker

so the point of that my msg is just for PA. since you said that PW is do 48khz resampling automatically by default. ok so Thanks you again, I have to adjust the text soon.

**Comment by NP-chaonay** [comment]
At now I have discontinued my old config to making new one that have better implementation instead. (which is better because using method from PulseEffects guide, but it diffcult and got more time to tweak even tweak it medium-level laziness)

While we waiting for it, however I just found that PulseEffect Convolver File from https://github.com/eedeidk/PulseAudio-IRSs/tree/master/IRS,

there is 1 file that similar to laptop state of SLS audio.

Ref and more info: https://github.com/NP-chaonay/Misc/discussions/2#discussioncomment-3127786

any good idea welcomed including the workaround like this.
